### DISCUSSION TITLE
How to configure Webpack for using a Worker

### DISCUSSION THREAD

**Discussion by webdiscus** [webdiscus/html-bundler-webpack-plugin#26] [Q&A]
Let's say I want to load a ServiceWorker. How would the webpack.common.js look and where would I save sw.js?

I've been getting an error trying to do this different ways, here saving sw.js in src where index.html is stored.

```
Failed to read response from Signed Web Bundle: 
Failed to read response: The Web Bundle does not contain a 
response for the provided URL: 
isolated-app://<ID>/home/user/telnet-client/src/sw.js
sw.js:1 
```

```html
<script type="module">
navigator.serviceWorker.register('./sw.js', {
  scope: './',
  type:'module'
});
</script>
```

**Comment by webdiscus** [comment]
@guest271314

To set up Webpack for both Web and Webworker, you need two different configurations because each configuration type must have its own `target` option.

For example, we have a html page and a Worker used on the page:

```js
const HtmlBundlerPlugin = require('html-bundler-webpack-plugin');

const webConfig = {
  target: 'web', // can be omitted, because default target is `web`
  plugins: [
    new HtmlBundlerPlugin({
      entry: {
        index: 'src/index.html',  // => dist/index.html
      },
      js: {
        inline: true,
      },
      css: {
        inline: true,
      },
    }),
  ],
}

const webWorkerConfig = {
  target: 'webworker', // <= very important set the target for workers
  entry: {
    // define workers here
    'workers/sw': './src/sw.js', // => dist/workers/sw.js
  }
}

module.exports = [webConfig, webWorkerConfig];
```

The Web Workers will be compiled and saved to the `dist/workers/` directory. I recommend using a separate directory for workers.

`src/index.html`

```html
<html>
<head>
  <meta charset="utf-8">
  <link href="./style.css" rel="stylesheet">
  <script src="./index.ts" defer="defer"></script>
</head>
<body>
  ...
  <script type="module">
     // the worker's path is a path relative by `dist/`
     navigator.serviceWorker.register('/workers/sw.js', {
       scope: './workers/', // <=  set to the directory where the service worker script is located, relative by `dist/`
       type: 'module'
    });
  </script>
</body>
</html>
```

Now you can use in your `src/sw.js` `import` and `require()`.

**Comment by guest271314** [comment]
I'm not sure where I place the files in the directory and what I write in the HTML to reference a file as `<script src="/path/to/file.js">` I keep getting an error.

**Reply by webdiscus** [reply]
in the `src` attribute must be the source file, relative to the html file:

```html
<script src="./relative-path/to/file.js">
```

**Reply by guest271314** [reply]
webpack.common.js
```
new HtmlBundlerPlugin({
      entry: {
        // define templates here
        index: 'src/index.html',
      },
      js: {
        // output filename of compiled JavaScript, used if inline is false
        filename: 'sw.js',
        inline: false, // inlines compiled JS into HTML using the <script> tag
      },
      css: {
        // output filename of extracted CSS, used if inline is false
        filename: 'css/[name].[contenthash:8].css',
        inline: true, // inlines CSS into HTML using the <style> tag
      },
    }),
```

src/index.html

```
  <!-- webpackIgnore: true -->
  <link rel="manifest" href="../assets/manifest.webmanifest">
  <script src="./sw.js" defer="defer"></script>
```

Error on Chromium in the IWA

```
Failed to read response from Signed Web Bundle: Failed to read response: 
The Web Bundle does not contain a response for the provided URL: isolated-app://<ID>/home/user/telnet-client/src/sw.js
sw.js:1 

Failed to load resource: the server responded with a status of 404 (Not Found) sw.js:1
```

What is the issue with the above?

I don't ordinarily use Webpack; that's how GoogleChromeLabs packaged their IWA/telnet-client demo. Ideally I could just create the HTML and JavaScript files to web bundle with something like https://github.com/paroga/cbor-js and sign the bundle without using Webpack.

**Reply by guest271314** [reply]
I'm not sure if a `ServiceWorker` will register on `isolated-app:` protocol. There are execptions to the `https:` protocol for `SrviceWorker`s: `chrome-extension:`. I created a Blob URL where a `Worker` did load.

**Reply by webdiscus** [reply]
can you please create an `issue` branch with your current code (with issue) and push in your [repo](https://github.com/guest271314/telnet-client).
Then I can try to help you.

**Reply by guest271314** [reply]
Sure. What I am testing locally looks something like https://github.com/guest271314/telnet-client/tree/issue.

**Reply by webdiscus** [reply]
don't use a fix filename, like ` filename: 'sw.js',`, otherwise you can get an error:
```
[webpack-dev-middleware] Error: Conflict: Multiple chunks emit assets to the same filename sw.js (chunks sw and hot-update)
```
Use a filename template, like `filename: '[name].js',` then the output filename will be the same as source file. `src/sw.js` => `dist/sw.js`

**Reply by webdiscus** [reply]
I don't see the error:
```
Failed to load resource: the server responded with a status of 404 (Not Found) sw.js:1
```

`npm start` and `npm run build` works

**Reply by webdiscus** [reply]
please push into `issue` branch you experiment code with Workers, where appeared the error.

**Reply by guest271314** [reply]
I posted the code where the errors appears.

**Reply by webdiscus** [reply]
src/index.html

```html
<html>
<head>
  <meta charset="utf-8">
  <!-- use references to the source files -->
  <link rel="manifest" href="../assets/manifest.webmanifest">
  <script src="./sw.js" defer="defer"></script>
</head>
<body>
</body>
</html>
```

src/sw.js
```js
console.log('Test');
```

webpack.common.js
```js
const path = require('path');
const HtmlBundlerPlugin = require('html-bundler-webpack-plugin');
const {CleanWebpackPlugin} = require('clean-webpack-plugin');
const CopyPlugin = require('copy-webpack-plugin');

module.exports = {
  module: {
    rules: [
      {
        test: /\.ts$/,
        use: 'ts-loader',
        exclude: /node_modules/,
      },
      {
        test: /\.css$/,
        use: [
          'css-loader',
        ],
      },
      // the bundler plugin allows webpack to copy the mainifest file
      // specified in the HTML from assets/ dir to dist/
      {
        test: /\.webmanifest$/,
        type: 'asset/resource',
        generator: {
          filename: '[name][ext]',
        },
      },
    ],
  },
  plugins: [
    new CleanWebpackPlugin(),
    new HtmlBundlerPlugin({
      entry: {
        // define templates here
        index: 'src/index.html',
      },
      js: {
        // output filename of compiled JavaScript, used if inline is false
        filename: 'sw.js',
        inline: false, // inlines compiled JS into HTML using the <script> tag
      },
      css: {
        // output filename of extracted CSS, used if inline is false
        filename: 'css/[name].[contenthash:8].css',
        inline: true, // inlines CSS into HTML using the <style> tag
      },
    }),
    new CopyPlugin({
      patterns: [
        // copy only images defined in the webmanifest file
        {from: 'assets/images', to: 'images'},
      ],
    }),
  ],
  resolve: {
    extensions: ['.ts', '.js'],
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: './',
    trustedTypes: {
      policyName: 'telnet#webpack',
    },
  },
};
```

It is this files?

Then `npm start` =>  ERROR in terminal `[webpack-dev-middleware] Error: Conflict: Multiple chunks emit assets to the same filename sw.js (chunks sw and hot-update)`, because you use wrong `js.filename`, plese set it as `filename: '[name].js'`


Then `npm run build` and open generated `dist/index.html` in FireFox, it works. I can see in the browser console `Test` output.

**Reply by guest271314** [reply]
Firefox doesn't implement Isolated Web Apps. How are you testing this in the browser?

**Reply by webdiscus** [reply]
## Testing with Webpack server

1. fix the `filename: '[name].js'` for js option
2. `npm start`
3. open `http://localhost:8080/` in chrome
4. open browser console:
```
[HMR] Waiting for update signal from WDS...
sw.js:1 Test
log.js:24 [HMR] Waiting for update signal from WDS...
index.js:551 [webpack-dev-server] Hot Module Replacement enabled.
index.js:551 [webpack-dev-server] Live Reloading enabled.
:8080/favicon.ico:1     Failed to load resource: the server responded with a status of 404 (Not Found)
index.js:551 [webpack-dev-server] Hot Module Replacement enabled.
index.js:551 [webpack-dev-server] Live Reloading enabled.
```

No errors.


## Testing generated `dist/`

1. `npm run build` => will be generated `dist/`
2.  open `dist/index.html` with chrome
3. browser console output:
```
Test
index.html:1 Access to internal resource at 'file:///Users/xxxxx/Developer/telnet-client/dist/manifest.webmanifest' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.
```

I don't understand you problem.

**Reply by guest271314** [reply]
I'm not using Webpack server. I'm trying to use Webpack as little as possible. I am using either a Deno, Node.js or txiki.js local TCP serverfor testing https://github.com/GoogleChromeLabs/telnet-client/pull/18. 

On Chromium/Chrome when I try to define a `<script>` with `src` set in `index.html` an error winds up being thrown.

Essesntially what I am trying to do is get around the CSP, COOP, COEP the IWA sets by default so I can use `postMessage()` to transfer the `ReadableStream` from `TCPSocket()` in the IWA to the arbotrary Web page, and write to the `WritableStream` from the Web page in the same way.

When testing I don't think the IWA is loading a `ServiceWorker` so I can intercept `fetch()` requests and remove the headers from the served response.

There's a few other ways to do this - even though the use case is supposedly not supposed; using an extension with `"externally_connectable"` in the IWA and Web page coordinated in the background MV3 `ServiceWorker`; and using Web API's alone to `open()` the IWA with SDP from a WebRTC `RTCPeerConnection` set as query string, and exchange SDP with the Web page using `<a> download` and `showOpenFilePicker()`. Both work. So I think I'm going to set the attempt to create a `ServiceWorker` for the IWA to intercept request aside right now; though I might just use `chrome.debugger` to intercept the request. For my use cases this is all about using Direct Sockets however I want to - even though Google folks have attempted to "isolate" that functionality, I have already created code to launch, control, and close the IWA from an arbtritrary Web page. Thanks for your help.

**Reply by webdiscus** [reply]
I can help to configure Webpack for generating a static HTML with assets, but CSP, COOP, COEP, IWA, MV3, RTCPeerConnection, showOpenFilePicker etc. it sounds like a foreign language to me.
I can't help for this, sorry.

**Reply by guest271314** [reply]
I didn't ask for help in those areas. I just wondered why the compiled IWA is throwing errors when I try to reference a compiled script in the HTML.

The only reason I was trying to load a `ServiceWorker` is to intecept the original request and remove certain headers from the response served with the `ServiceWorker`. However, it doesn't look like the IWA can actually load a `ServiceWorker` on `isolated-app:` protocol.

> it sounds like a foreign language to me.

Same thing for Webpack as to me. I really have no use for Webpack. Webpack just happens to be used by the IWA folks at GoogleChromeLabs, that's why I filed this issue [Create an example without depending on TypeScript, Webpack, Node.js, npm #12](https://github.com/GoogleChromeLabs/telnet-client/issues/12). Thanks again, cheers.

**Reply by guest271314** [reply]
I spent the past day testing connecting an arbitrary Web page to the `isolated-app:` page using WebRTC data channels. I successfully achieved that. Now it turns out the inline script doesn't expose `TCPSocket`, so I'm going to have to figure out how to load the script using `src`.

```
VM136:1 Uncaught DOMException: Failed to construct 'TCPSocket': Frame is not sufficiently isolated to use Direct Sockets.
```

**Reply by webdiscus** [reply]
you cannot load a Worker script directly into `<script src="...">`. It won't work.

Firstly you should create a normal JS file (e.g. 'app.js') and load this file in html `<script src="./path/to/app.js">`.
Then, your Worker JS file must be loaded in this `app.js`:

_app.js_
```js
const workerUrl = '/workers/myWorker.js'; // <= this is the output filename relative to the `dist/`
const worker = new Worker(workerUrl);
// ..
```

The `src/js/myWorker.js` file should be compiled via Webpack if this script uses `import` or `require`. If your `myWorker.js` file works in Chromium w/o compilation, then simple place it directly in the `dist/` directory, w/o compilation.

**Reply by guest271314** [reply]
I'm not talking about a `Worker` script. I'm just talking about using your pluging to load a non-Worker script, which I havn't been able to do yet without errors being thrown.

**Reply by guest271314** [reply]
I think what was throwing were query string parameters passed as Isolated Web App URL to `open()`. I tried with the source code from the telnet-client repository before your forked PR.

**Reply by guest271314** [reply]
FWIW This is a summary of what I wound up doing for this particular version of controlling an IWA from an arbitrary Web page using the the Webpack/Node.js/et al. code from https://github.com/guest271314/telnet-client/tree/guest271314-patch-1:

- `open('isolated-app://<ID>?sdp=<btoa(SDP)>')` where "SDP" is Session Description Protocol from a WebRTC RTCPeerConnection which creates a WebRTC Data Channel to the IWA (by using WICG File System Access API to read the SDP the WebRTC `RTCPeerConnection` generates and writes to a local file)
- `"externally_connectable"` to connect the IWA to MV3 `ServiceWorker` extension on Chromium to launch a local txiki.js TCP server with Native Messaging, whereafter I send messages from the arbitrary Web page to the WebRTC Data Channel in the IWA which writes the message to the writable side of the TCPSocket and sends the result from the TCP socket to the arbitrary Web page.
